compiler/src/dotty/tools/repl/ReplDriver.scala

```diff
@@ -209,20 +209,24 @@ class ReplDriver(settings: Array[String],
         /* complete = */ false  // if true adds space when completing
       )
     }
-    implicit val state = newRun(state0)
-    compiler
-      .typeCheck(expr, errorsAllowed = true)
-      .map { tree =>
-        val file = SourceFile.virtual("<completions>", expr, maybeIncomplete = true)
-        val unit = CompilationUnit(file)(using state.context)
-        unit.tpdTree = tree
-        given Context = state.context.fresh.setCompilationUnit(unit)
-        val srcPos = SourcePosition(file, Span(cursor))
-        val (_, completions) = Completion.completions(srcPos)
-        completions.map(_.label).distinct.map(makeCandidate)
-      }
-      .getOrElse(Nil)
-  }
+
+    if expr.startsWith(":") then
+      ParseResult.commands.map(command => makeCandidate(command._1))
```

Comment:
I assume we should also filter by the prefix after ":". It's not only for consistency but then also one could type e.g. `:s<TAB>` and then `:settings` would be completed as there are no other options with the same prefix.

Reply:
Ah of course, I'll add this.

Reply:
Alright, added this in although the bahavior of jline here seems a bit unexpected where if you give it `:do` and you offer `:doc` as a completion it will result in `::doc` in the repl. So int he case where it's only `:` we include the `:` in the completion, but in the other cases like `:do` we don't and offer back `doc` as a completion. While this looks funny, this results in the behavior you actually want in the REPL.

![2022-03-07 10 36 15](https://user-images.githubusercontent.com/13974112/157005852-f7977d99-6d93-4387-a97c-e46228f7d340.gif)

Reply:
I think you can make JLine behave as needed here, as evidenced by Scala 2 not having the same issue.

I'm not sure of all the details, but in Scala 2 we have

```
case class CompletionResult(line: String, cursor: Int, ...)
```

and when completing a colon-command, `cursor` is 1 instead of 0

Reply:
Yea I noticed @prolativ also added some helpful info about this down below. I'll work on cleaning this up.